File: data_structures/c++/singly_linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

struct node
{
        int data;
        node *next;
};
struct node *head = NULL;

void display()
{
        struct node *temp;
        if (head == NULL)
        {
                cout << "list is empty\n\n";
        }
        else
        {
                temp = head;
                cout << "Singly linked list: ";
                while (temp != NULL)
                {
                        cout << temp->data << " ";
                        temp = temp->next;
                }
        }
        cout << "\n\n";
}
int temp_reverse(struct node *temp)
{
        if (temp == NULL)
        {
                cout << "reverse: ";
                return 0;
        }
        temp_reverse(temp->next);
        cout << temp->data << " ";
}
void insert_beg(int x)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));

        if (head == NULL)
        {
                new_node->data = x;
                new_node->next = NULL;
                head = new_node;
        }
        else
        {
                new_node->data = x;
                new_node->next = head;
                head = new_node;
        }
        display();
}
void insert_end(int x)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL)
        {
                new_node->data = x;
                new_node->next = NULL;
                head = new_node;
        }
        else
        {
                new_node->data = x;
                new_node->next = NULL;

                struct node *temp = head;

                while (temp->next != NULL)
                {
                        temp = temp->next;
                }
                temp->next = new_node;
        }
        display();
}
void insert_at_n(int x, int n)
{
        struct node *new_node = (struct node *)malloc(sizeof(struct node));
        if (head == NULL && n == 0)
        {
                new_node->data = x;
                new_node->next = NULL;
                head = new_node;
        }
        else if (head == NULL && n > 0)
        {
                cout << "Invalid Index" << endl;
        }
        else if (n == 0)
        {
                insert_beg(x);
        }
        else
        {
                struct node *temp = head;
                bool valid = true;

                while (--n)
                {
                        temp = temp->next;
                        if (temp == NULL)
                        {
                                cout << "Invalid Index" << endl;
                                valid = false;
                                break;
                        }
                }
                if (valid)
                {
                        new_node->data = x;

                        new_node->next = temp->next;
                        temp->next = new_node;
                }
                display();
        }
}
void delete_from_beg()
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                head = head->next;
                temp->next = NULL;
        }
        display();
}
void delete_from_end()
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                while (temp->next->next != NULL)
                {
                        temp = temp->next;
                }
                temp->next = NULL;
        }
        display();
}
void delete_from_n(int n)
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else if (n == 0)
        {
                struct node *temp = head;
                head = head->next;
                temp->next = NULL;
        }
        else
        {
                struct node *temp = head;
                bool valid = true;
                while (--n)
                {
                        temp = temp->next;
                        if (temp == NULL)
                        {
                                cout << "Invalid Index" << endl;
                                valid = false;
                                break;
                        }
                }
                if (valid)
                {
                        temp->next = temp->next->next;
                }
        }
        display();
}
void delete_the_element(int x)
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                bool found = false;
                int n = 0;
                struct node *temp = head;
                while (temp != NULL)
                {
                        if (temp->data == x)
                        {
                                delete_from_n(n);
                                n--;
                                found = true;
                        }
                        temp = temp->next;
                        n++;
                }
                if (!found)
                {
                        cout << "The element " << x << " is not present in list" << endl;
                }
        }
}
void reverse_display()
{
        if (head == NULL)
        {
                cout << "List is already empty!" << endl;
        }
        else
        {
                struct node *temp = head;
                temp_reverse(temp);
                cout << endl;
        }
}

void custom_display()
{
        struct node *temp;
        if (head == NULL)
        {
                cout << "list is empty\n\n";
        }
        else
        {
                temp = head;
                cout << "|" << head << "|-->";
                while (temp != NULL)
                {
                        cout << "|" << temp->data << "-" << temp->next << "|";
                        if (temp->next != NULL)
                        {
                                cout << "-->";
                        }
                        temp = temp->next;
                }
        }
        cout << "\n\n";
}

int main()
{
        insert_beg(2);
        insert_beg(7);
        insert_beg(5);
        insert_beg(4);
        insert_end(45);
        insert_at_n(78, 4);
        delete_from_end();
        delete_from_n(2);
        delete_from_beg();
        reverse_display();
        custom_display();
}

File: data_structures/python/singly_linked_list.py
import sys
stdwrite = lambda line: sys.stdout.write(line) 

class SinglyLinkedList:
    class __Node:
        def __init__(self):
            self.data = None
            self.next = None

        def set_data(self, data):
            self.data = data

        def get_data(self):
            return self.data

        def set_next(self, next):
            self.next = next

        def get_next(self):
            return self.next
    __head = None

    def insert_at_beg(self, data):
        if not SinglyLinkedList.__head:
            new = self.__Node()
            new.set_data(data)
            SinglyLinkedList.__head = new
        else:
            new = self.__Node()
            new.set_data(data)
            new.set_next(SinglyLinkedList.__head)
            SinglyLinkedList.__head = new
        self.display()

    def insert_at_end(self, data):
        if not SinglyLinkedList.__head:
            new = self.__Node()
            new.set_data(data)
            SinglyLinkedList.__head = new
        else:
            new = self.__Node()
            new.set_data(data)
            temp = SinglyLinkedList.__head
            while temp.get_next():
                temp = temp.get_next()
            temp.set_next(new)
        self.display()

    def insert_at_n(self, data, n):
        if not SinglyLinkedList.__head:
            stdwrite("List is not create yet\n")
            return False
        elif n==0:
            self.insert_at_beg(data)
        else:
            counter = 0
            temp = SinglyLinkedList.__head
            while temp:
                if counter==n-1:
                    if not temp.get_next():
                        self.insert_at_end(data)
                        return True
                    new = self.__Node()
                    new.set_data(data)
                    new.set_next(temp.get_next())
                    temp.set_next(new)
                    self.display()
                    return True
                temp = temp.get_next()
                counter += 1
            stdwrite("This is index does not exist in the list\n")
            return False

    def delete_from_beg(self):
        if not SinglyLinkedList.__head:
            stdwrite("List is already empty\n")
        else:
            SinglyLinkedList.__head = SinglyLinkedList.__head.get_next()
            self.display()
    def delete_from_end(self):
        if not SinglyLinkedList.__head:
            stdwrite("List is already empty\n")
        else:
            temp = SinglyLinkedList.__head
            if not temp.get_next():
                SinglyLinkedList.__head = None
            else:
                while temp.get_next().get_next():
                    temp = temp.get_next()
                temp.set_next(None)
                self.display()

    def delete_from_n(self, n):
        if not SinglyLinkedList.__head:
            stdwrite("List is already empty\n")
        elif n==0:
            self.delete_from_beg()
        else:
            temp = SinglyLinkedList.__head
            counter = 0 
            while temp:
                if counter == n-1:
                    try: 
                        temp.set_next(temp.get_next().get_next())
                        self.display()
                        return True
                    except:
                        stdwrite("This position does not exist\n")
                        return False
                temp = temp.get_next()
                counter += 1
            stdwrite("This position does not exist\n")

    def display(self):
        if not SinglyLinkedList.__head:
            stdwrite("List is empty\n")
        else:
            stdwrite("Singly Linked List: head")
            temp = SinglyLinkedList.__head
            while temp:
                stdwrite(" -> "+str(temp.get_data()))
                temp = temp.get_next()
            stdwrite("\n")
    def __reverse(self, temp):
        if temp==None:
            return 
        self.__reverse(temp.get_next())
        stdwrite("<- "+str(temp.get_data())+" ")

    def print_reverse(self):
        if not SinglyLinkedList.__head:
            stdwrite("The list is empty\n")
        else:
            stdwrite("Reverse Linked List: ")
            temp = SinglyLinkedList.__head
            self.__reverse(temp)
            stdwrite("<- head")

if __name__ == '__main__':
    li = SinglyLinkedList()
    li.insert_at_beg(4)
    li.insert_at_beg(5)
    li.insert_at_beg(41)
    li.insert_at_end(7)
    li.insert_at_n(10, 2)
    li.delete_from_beg()
    li.delete_from_end()
    li.delete_from_n(2)
    li.print_reverse()
